legate/core/runtime.py

```diff
@@ -711,14 +729,389 @@ def record_partition(self, index_space, functor, index_partition):
+            beta = self.constraints[2].apply2(self.contexts, self.runtime, self.ops, beta[1], self.partitioners, self.strategies)
+        #print("beta", beta)
+        """
+        stoptime = datetime.datetime.now()
+        delta=stoptime-starttime
+        total = delta.total_seconds() * 1000.0
+        print("applytime", total)
+        """
+        starttime = datetime.datetime.now()
+        #drint("fuse results", results)
+  
+        #intersect intervals
+        #this is an inefficent way of doing this,
+        #"""
+        if not streamApply:
+            curr_set = interval_sets[0]
+            for interval_set in interval_sets[1:]:
+                newset = []
+                for aset in curr_set:
+                    for bset in interval_set:
+                        if not (aset[0] > bset[1] or bset[0] > aset[1]): 
+                            news = max(aset[0], bset[0])
+                            newe = min(aset[1], bset[1])
+                            newset.append((news, newe))
+                curr_set=newset
+        #"""
+            fusable,final_set = self.supress_small_fusions(curr_set, self.runtime._fusion_threshold)
+        else:
+            fusable,final_set = self.supress_small_fusions(beta[1], self.runtime._fusion_threshold)
+         
+        """
+        stoptime = datetime.datetime.now()
+        delta=stoptime-starttime
+        total = delta.total_seconds() * 1000.0
+        print("filtertime", total)
+        """
+
+        #drint("curset", curr_set)
+
+        #drint("final_set", final_set)
+        #drint("all fusable", fusable)
+        #drint("intervals", interval_sets)
+        #return reduce(lambda x,y: x and y, all_fusable), final_set, self.strategies
+        return fusable, final_set, self.strategies
+
+class FusionConstraint(object):
+    def apply(self, contexts, runtime, ops, partitioners, strategies):
+        """"
+         Abstract class for determining a rule that constrains
+         which legate operations can be fused
+         """
+        raise NotImplementedError("Implement in derived classes")
+
+
+class NumpyContextExists(FusionConstraint):
+    def apply(self, contexts, runtime, ops, partitioners, strategies):
+        if "cunumeric" in contexts:
+            return True, [(0, len(ops))]
+        else:
+           return False, [(0,0)]
+
+
+class AllValidOps(FusionConstraint):
+    """
+    Class for only fusing only potentially fusable ops.
+    This class performs the first pass of legality filtering
+    """
+    def __init__(self):
+        self.validIDs = set()
+        self.terminals = set()
+        #these ops are almost always fusable
+        self.validIDs.add(2) #Binary op
+        self.validIDs.add(10) #Fill op
+        self.validIDs.add(20) #Where op
+        #self.validIDs.add(6) #Where op
+        #self.validIDs.add(5) #Convert op
+        #self.validIDs.add(7) #dot op
+        #self.terminals.add(7) #dot op is only fusable as a terminal op in a window
+
```

Comment:
I initially introduced the notion of a 'terminal' op that is fusable only at the end of a window. I don't use this so will remove it.